sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JsonUtils.java

```diff
@@ -308,6 +303,79 @@ public static <T> T getNullableProperty(JsonReader jsonReader, Function<JsonRead
+     * An {@link IllegalStateException} will be thrown if the {@link JsonReader#currentToken()} isn't
+     * {@link JsonToken#START_OBJECT} or {@link JsonToken#NULL} when the method is called.
+     *
+     * @param jsonReader The {@link JsonReader} being read.
+     * @param fieldNameConsumer The field name consumer function.
+     * @throws IllegalStateException If {@link JsonReader#currentToken()} isn't {@link JsonToken#START_OBJECT} or
+     * {@link JsonToken#NULL} when this method is called.
+     */
+    public static void readFields(JsonReader jsonReader, Consumer<String> fieldNameConsumer) {
+        readFields(jsonReader, false, fieldName -> {
+            fieldNameConsumer.accept(fieldName);
+            return false;
+        });
+    }
+
+    /**
+     * Reads the fields of a JSON object until the end of the object is reached.
+     * <p>
+     * The passed {@link JsonReader} will point to the field value each time {@code fieldNameConsumer} is called.
+     * <p>
+     * An {@link IllegalStateException} will be thrown if the {@link JsonReader#currentToken()} isn't
+     * {@link JsonToken#START_OBJECT} or {@link JsonToken#NULL} when the method is called.
+     * <p>
+     * If {@code readAdditionalProperties} is true and {@code fieldNameConsumer} returns false the JSON field value
+     * will be read as if it were an additional property. After the object completes reading the untyped additional
+     * properties mapping will be returned, this may be null if there were no additional properties in the JSON object.
+     *
+     * @param jsonReader The {@link JsonReader} being read.
+     * @param readAdditionalProperties Whether additional properties should be read.
+     * @param fieldNameConsumer The field name consumer function.
+     * @return The additional property map if {@code readAdditionalProperties} is true and there were additional
+     * properties in the JSON object, otherwise null.
+     * @throws IllegalStateException If {@link JsonReader#currentToken()} isn't {@link JsonToken#START_OBJECT} or
+     * {@link JsonToken#NULL} when this method is called.
+     */
+    public static Map<String, Object> readFields(JsonReader jsonReader, boolean readAdditionalProperties,
+        Function<String, Boolean> fieldNameConsumer) {
+        if (jsonReader.currentToken() != JsonToken.START_OBJECT && jsonReader.currentToken() != JsonToken.NULL) {
+            throw new IllegalStateException("Expected the current token of the JsonReader to either be "
+                + "START_OBJECT or NULL. It was: " + jsonReader.currentToken());
+        }
+
+        if (jsonReader.currentToken() == JsonToken.NULL) {
+            return null;
+        }
+
+        Map<String, Object> additionalProperties = null;
+
+        while (jsonReader.nextToken() != JsonToken.END_OBJECT) {
+            String fieldName = jsonReader.getFieldName();
+            jsonReader.nextToken();
+
+            boolean consumed = fieldNameConsumer.apply(fieldName);
```

Comment:
If I understand correctly, it assumes that the consumer method works with the same instance of `reader` as this method, which is not so straightforward. Could consumer have `reader` as an argument?